File: app/src/main/java/com/example/voice_recorder/MainActivity.java
package com.example.voice_recorder;

import java.io.File;
import java.util.List;
import android.Manifest;
import android.os.Bundle;
import com.vk.api.sdk.VK;
import java.util.ArrayList;
import android.widget.Toast;
import android.os.Environment;
import android.content.Intent;
import android.widget.ImageView;
import com.vk.api.sdk.auth.VKScope;
import androidx.annotation.NonNull;
import android.content.ContextWrapper;
import androidx.core.app.ActivityCompat;
import com.vk.api.sdk.auth.VKAccessToken;
import android.content.pm.PackageManager;
import com.vk.api.sdk.auth.VKAuthCallback;
import androidx.core.content.ContextCompat;
import androidx.appcompat.app.AppCompatActivity;
import androidx.recyclerview.widget.RecyclerView;
import com.example.voice_recorder.adapters.RecyclerAdapter;

public class MainActivity extends AppCompatActivity {

    final int MICROPHONE_PERMISSION = 100;
    final int WRITE_PERMISSION = 200;
    final int READ_PERMISSION = 300;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        ArrayList<VKScope> vkDocuments = new ArrayList<>();
        vkDocuments.add(VKScope.DOCS);
        VK.login(this, vkDocuments);
    }

    //Wait until permissions will be granted
    @Override
    protected void onResume() {
        super.onResume();
        if(hasMicrophone() && allPermissionsGranted()){
            init();
        }
    }

    @Override
    protected void onActivityResult(int requestCode, int resultCode, Intent data) {
        VKAuthCallback callback = new VKAuthCallback(){
            @Override
            public void onLoginFailed(int i) {
                Toast.makeText(getApplicationContext(), "Authorisation failed", Toast.LENGTH_LONG).show();
            }

            @Override
            public void onLogin(@NonNull VKAccessToken vkAccessToken) {

            }
        };

        if(data == null || !VK.onActivityResult(requestCode, resultCode, data, callback))
            super.onActivityResult(requestCode, resultCode, data);
    }

    //Initialization
    private void init(){
        List<RecordCard> recordCards = new ArrayList<>();

        File directory = getDirectory();
        File[] files = directory.listFiles();

        ImageView recordButton = (ImageView) findViewById(R.id.recordButton);
        RecyclerView recycler = (RecyclerView) findViewById(R.id.recyclerView);


        AudioRecorder audioRecorder = new AudioRecorder(
                getResources(), directory.getPath(), recordCards
        );
        audioRecorder.loadData(files, recordCards);

        RecyclerAdapter adapter = new RecyclerAdapter(recordCards);
        recycler.setAdapter(adapter);
        audioRecorder.setRecordButton(recordButton, recordCards, adapter);
    }

    //Get directory with files
    private File getDirectory(){
        ContextWrapper contextWrapper = new ContextWrapper(getApplicationContext());
        return contextWrapper.getExternalFilesDir(Environment.DIRECTORY_MUSIC);
    }

    //Check microphone is unable
    private boolean hasMicrophone(){
        return this.getPackageManager().hasSystemFeature(PackageManager.FEATURE_MICROPHONE);
    }

    //Get record audio permission
    private boolean getRecordPermission(){
        if(ContextCompat.checkSelfPermission(this, android.Manifest.permission.RECORD_AUDIO)
                == PackageManager.PERMISSION_DENIED){
            ActivityCompat.requestPermissions(this, new String[] {
                    android.Manifest.permission.RECORD_AUDIO}, MICROPHONE_PERMISSION);
            return false;
        }
        return true;
    }

    //Get write external storage permission
    private boolean getWritePermission(){
        if(ContextCompat.checkSelfPermission(this, Manifest.permission.WRITE_EXTERNAL_STORAGE)
                == PackageManager.PERMISSION_DENIED){
            ActivityCompat.requestPermissions(this, new String[] {
                    Manifest.permission.WRITE_EXTERNAL_STORAGE}, WRITE_PERMISSION);
            return false;
        }
        return true;
    }

    //Get read external storage permission
    private boolean getReadPermission(){
        if(ContextCompat.checkSelfPermission(this, Manifest.permission.READ_EXTERNAL_STORAGE)
                == PackageManager.PERMISSION_DENIED){
            ActivityCompat.requestPermissions(this, new String[] {
                    Manifest.permission.WRITE_EXTERNAL_STORAGE}, READ_PERMISSION);
            return false;
        }
        else return true;
    }

    private boolean allPermissionsGranted(){
        if(!getRecordPermission()) return false;
        if(!getWritePermission()) return false;
        return getReadPermission();
    }

}

File: app/src/main/java/com/michel/dictophone/MainActivity.kt
package com.michel.dictophone

import android.content.ContextWrapper
import android.graphics.Canvas
import android.media.MediaRouter.SimpleCallback
import android.os.Bundle
import android.os.Environment
import android.util.Log
import android.view.View
import android.widget.ImageView
import android.widget.TextView
import androidx.appcompat.app.AppCompatActivity
import androidx.recyclerview.widget.ItemTouchHelper
import androidx.recyclerview.widget.RecyclerView
import com.michel.dictophone.adapters.RecyclerAdapter
import it.xabaras.android.recyclerview.swipedecorator.RecyclerViewSwipeDecorator
import java.io.File

class MainActivity : AppCompatActivity() {

    private val utils = Utils()
    private var permissionWasGranted = false
    private val recordCards: MutableList<RecordCard> = ArrayList()

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)
    }

    override fun onResume() {
        super.onResume()
        if(!permissionWasGranted && utils.readPermissionGranted(this)) {
            permissionWasGranted = true
            init()
            Log.w("Resume", "Resumed")
        }
    }

    //Initialization
    private fun init() {
        val directory = getDirectory()
        val files = directory!!.listFiles()
        val recordButton = findViewById<View>(R.id.recordButton) as ImageView
        val durationTextView = findViewById<View>(R.id.durationText) as TextView
        val recycler = findViewById<View>(R.id.recyclerView) as RecyclerView
        val recorder = Recorder(resources, directory.path)
        recorder.loadData(files, recordCards)
        val adapter = RecyclerAdapter(recordCards, recorder, this)
        recycler.adapter = adapter

        val itemTouchHelper = ItemTouchHelper(getCallBack(adapter))
        itemTouchHelper.attachToRecyclerView(recycler)

        recorder.setRecordButton(recordButton, durationTextView, recordCards, adapter, this)
    }


    //Returns directory with files
    private fun getDirectory(): File? {
        val contextWrapper = ContextWrapper(applicationContext)
        return contextWrapper.getExternalFilesDir(Environment.DIRECTORY_MUSIC)
    }

    private fun getCallBack(adapter: RecyclerAdapter) : ItemTouchHelper.SimpleCallback {
        return object : ItemTouchHelper.SimpleCallback(
            0,
            ItemTouchHelper.LEFT
        ) {
            override fun onMove(
                recyclerView: RecyclerView,
                viewHolder: RecyclerView.ViewHolder,
                target: RecyclerView.ViewHolder
            ): Boolean {
                TODO("Not yet implemented")
            }

            override fun onSwiped(viewHolder: RecyclerView.ViewHolder, direction: Int) {
                val position = viewHolder.adapterPosition
                recordCards[position].deleteRecord()
                recordCards.removeAt(position)
                adapter.notifyItemRemoved(position)
            }

            override fun onChildDraw(
                c: Canvas,
                recyclerView: RecyclerView,
                viewHolder: RecyclerView.ViewHolder,
                dX: Float,
                dY: Float,
                actionState: Int,
                isCurrentlyActive: Boolean
            ) {

                RecyclerViewSwipeDecorator.Builder(
                    c,
                    recyclerView,
                    viewHolder,
                    dX,
                    dY,
                    actionState,
                    isCurrentlyActive
                ).addSwipeLeftActionIcon(R.drawable.delete)
                    .setActionIconTint(resources.getColor(R.color.red))
                    .create()
                    .decorate()

                super.onChildDraw(
                    c,
                    recyclerView,
                    viewHolder,
                    dX,
                    dY,
                    actionState,
                    isCurrentlyActive
                )
            }

        }
    }
}